### PR TITLE
virtualbox-image module: autoresize root

### PR DESCRIPTION
**Pull request by edef1c**
This goes together with NixOS/nixops#389.
/cc @peti

### CODE REVIEW

**Comment by nathan7** [comment]
It's worth noting that `nixos/modules/virtualisation/virtualbox-grow-partition.nix` is pretty much a copy of `nixos/modules/virtualisation/amazon-grow-partition.nix`, but `s/xvda/sda/g`. Maybe we should abstract that, since it seems like a common enough task. (for example, I'd like to use it for my DigitalOcean images too)

**Comment by peti** [comment]
:+1:

**Comment by fpletz** [comment]
Yes, please abstract this! Thanks!

**Comment by dezgeg** [comment]
I suppose you could modify the existing script to whether the root device is `xvda` or `sda`, using logic like this: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/cd-dvd/sd-image.nix#L108

**Comment by nathan7** [comment]
@dezgeg Though that would work, I'd rather avoid adding more auto-detection magic to the current code path — maybe using `fileSystems."/".device` instead of full autodetection would do the trick, though.

**Comment by dezgeg** [comment]
Yes, looking at `fileSystems."/".device` was my idea as well.

**Comment by nathan7** [comment]
I've hit a snag: making the inclusion of the new grow-image module conditional on `config.ec2.hvm` for the EC2 image causes an infinite recursion. I'm guessing that's unavoidable without adding an option, but  I'm unsure what it should be called.

**Comment by dezgeg** [comment]
Maybe some clever use of `mkIf` could avoid the infinite recursion, but not totally sure on that...

**Comment by nathan7** [comment]
@dezgeg mkIf just pushes the `if` down into individual attributes for attribute sets, which doesn't really apply here. Conditional imports are circular no matter what, afaict.

**Comment by nathan7** [comment]
I've changed the code to use an option with the straw name `bob`, until I come up with something better. Seems to work fine [=

**Comment by domenkozar** [comment]
ready? :)

File: nixos/modules/virtualisation/grow-partition.nix [hunk 1]
```diff
@@ -28,19 +28,32 @@ in
...
+  options = {
+
+    bob = mkOption {
+      type = types.bool;
+      default = false;
+    };
+
+  };
+
+  config = {
```
Comment (domenkozar) [inline nixos/modules/virtualisation/grow-partition.nix]: This should check if `bob` is set

File: nixos/modules/virtualisation/grow-partition.nix [hunk 2]
```diff
@@ -28,19 +28,32 @@ in
 
 {
 
-  config = mkIf config.ec2.hvm {
+  options = {
+
+    bob = mkOption {
```
Comment (domenkozar) [inline nixos/modules/virtualisation/grow-partition.nix]: Why `bob` as a name?

Comment (nathan7) [inline nixos/modules/virtualisation/grow-partition.nix]: Because I have no idea what to name it! Got a suggestion?

**Comment by domenkozar** [comment]
@nathan7:

I'd like to merge this, but a few more issues first:

- rename `bob` to something like `virtualisation.growPartition`
- update comment at the top of `nixos/modules/virtualisation/grow-partition.nix` to reflect new state.

**Comment by nathan7** [comment]
@domenkozar How's this?

**Comment by eqyiel** [comment]
Hey, I think you need to add `with lib` or prefix `mkIf`, `mkOption` and `types` with `lib.`!

**Comment by nathan7** [comment]
@eqyiel eep, sorry, I messed that up while rebasing. thanks for fixing that :heart:

**Pull request merged**